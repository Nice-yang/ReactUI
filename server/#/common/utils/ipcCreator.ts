import * as Electron from 'electron';
import * as ipc from 'electron-better-ipc';

export function mainToRendererRequest<T, P>(channel: string) {
  const callRenderer = (win: Electron.BrowserWindow, arg: T) =>
    ipc.callRenderer && ipc.callRenderer(win, channel, arg);
  const answerMain = (cb: (arg: T) => Promise<P>) =>
    ipc.answerMain && ipc.answerMain(channel, cb);

  return {
    callRenderer,
    answerMain,
  };
}
