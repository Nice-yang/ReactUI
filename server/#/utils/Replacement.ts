import * as stable from 'stable';
import * as ts from 'typescript';

export class Replacement {
  public static insert(pos: number, text: string) {
    return new Replacement(pos, pos, text);
  }

  public static delete(start: number, end: number) {
    return new Replacement(start, end, '');
  }

  public static replaceNodeWithText<T extends ts.Node, Y extends ts.Node>(
    source: T,
    node: Y,
    text: string,
  ) {
    return [
      Replacement.delete(
        node.getStart() - source.getStart(),
        node.getEnd() - source.getStart(),
      ),
      Replacement.insert(node.getStart() - source.getStart(), text),
    ];
  }

  public static applyReplacements(source: string, replacements: Replacement[]) {
    replacements = stable(replacements, (r1, r2) => r2.start - r1.start);

    for (const replacement of replacements) {
      source =
        source.slice(0, replacement.start) +
        replacement.text +
        source.slice(replacement.end);
    }

    return source;
  }

  constructor(
    readonly start: number,
    readonly end: number,
    readonly text = '',
  ) {}
}
