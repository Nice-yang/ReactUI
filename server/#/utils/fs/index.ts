import * as fs from 'fs';
import { bindNodeCallback } from 'rxjs';

export const writeFile = bindNodeCallback(fs.writeFile);
