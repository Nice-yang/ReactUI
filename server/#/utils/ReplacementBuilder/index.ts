import { Replacement } from '#/utils/Replacement';
import * as ts from 'typescript';

export class ReplacementBuilder {
  private replacements: Replacement[];
  private sourceFile: ts.SourceFile;

  constructor(sourceFile: ts.SourceFile, replacements: Replacement[] = []) {
    this.sourceFile = sourceFile;
    this.replacements = replacements;
  }

  public insert(pos: number, text: string) {
    this.replacements.push(Replacement.insert(pos, text));
  }

  public delete(start: number, end: number) {
    this.replacements.push(Replacement.delete(start, end));
  }

  public deleteNode(node: ts.Node) {
    this.replacements.push(Replacement.delete(node.getStart(), node.getEnd()));
  }

  public replaceNodeWithText(node: ts.Node, text: string) {
    this.replacements = this.replacements.concat([
      Replacement.delete(node.getStart(), node.getEnd()),
      Replacement.insert(node.getStart(), text),
    ]);
  }

  public applyReplacements() {
    return Replacement.applyReplacements(
      this.sourceFile.getText(),
      this.replacements,
    );
  }
}
