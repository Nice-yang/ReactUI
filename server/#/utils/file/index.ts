import { Fault } from '#/common/models/Fault';
import { readFile, writeFile } from 'fs';
import * as R from 'ramda';
import { bindNodeCallback, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';

export const readFile$ = bindNodeCallback(readFile);
export const writeFile$ = bindNodeCallback(writeFile);
export const readFileToString$ = (fileName: string) =>
  readFile$(fileName).pipe(
    catchError(err =>
      throwError(
        new Fault('Cannot read file', { error: JSON.stringify(err), fileName }),
      ),
    ),
    map(R.toString),
  );
