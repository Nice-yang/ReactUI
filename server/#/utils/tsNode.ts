import { IJSXSource } from '#/common/models/file';
import * as ts from 'typescript';

export const getInfo = (node: ts.Node) => {
  const file = node.getSourceFile();

  const { line, character: column } = file.getLineAndCharacterOfPosition(
    node.getStart(),
  );

  const fileName = file.fileName;
  const lineNumber = line + 1;
  const columnNumber = column + 1;

  return {
    fileName,
    lineNumber,
    columnNumber,
  };
};

export const isAncestor = (a?: ts.Node, b?: ts.Node): boolean => {
  if (!b || !a) {
    return false;
  }

  if (b.parent === a) {
    return true;
  }

  return isAncestor(a, b.parent);
};

export const findNode = <T extends ts.Node>(
  rootNode: ts.Node,
  target?: IJSXSource,
  predicate?: (node: ts.Node) => boolean,
): T | undefined => {
  let resolvedNode;

  function _traverse(node: ts.Node) {
    const { lineNumber, columnNumber } = getInfo(node);

    if (
      (target &&
        (lineNumber === target.lineNumber &&
          columnNumber === target.columnNumber)) ||
      (predicate && predicate(node))
    ) {
      resolvedNode = node;
      return;
    }

    ts.forEachChild(node, _traverse);
  }

  _traverse(rootNode);

  return resolvedNode;
};

export const findAncestorNode = <T extends ts.Node>(
  rootNode: T,
  predicate: (node: ts.Node) => boolean,
): T | undefined => {
  let resolvedNode;

  let node = rootNode.parent;

  while (node) {
    if (predicate(node)) {
      resolvedNode = node as T;
      break;
    }

    node = node.parent;
  }

  return resolvedNode;
};

export const findAllNodes = <T extends ts.Node>(
  rootNode: ts.Node,
  predicate: (node: ts.Node) => boolean,
): T[] => {
  const resolvedNodes: T[] = [];

  function _traverse(node: ts.Node) {
    if (predicate(node)) {
      resolvedNodes.push(node as T);
    }

    ts.forEachChild(node, _traverse);
  }

  _traverse(rootNode);

  return resolvedNodes;
};
