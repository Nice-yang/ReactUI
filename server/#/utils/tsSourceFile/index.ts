import { readFileToString$ } from '#/utils/file';
import * as _ from 'lodash';
import { map } from 'rxjs/operators';
import * as ts from 'typescript';

export const createSourceFileFromText = (fileName: string, text: string) => {
  const file = ts.createSourceFile(
    fileName,
    text,
    ts.ScriptTarget.Latest,
    /* setParentNodes */ true,
    _.endsWith(fileName, 'tsx') ? ts.ScriptKind.TSX : undefined,
  );

  return file;
};

export const createSourceFile$ = (fileName: string) =>
  readFileToString$(fileName).pipe(
    map(text => ({ file: createSourceFileFromText(fileName, text), text })),
  );
