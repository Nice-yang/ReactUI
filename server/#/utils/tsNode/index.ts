import { Fault } from '#/common/models/Fault';
import { TagCursor } from '#/common/models/file';
import { catchThrowFault } from '#/operators/catchThrowError';
import * as R from 'ramda';
import { Observable } from 'rxjs';
import { pluck, switchMap } from 'rxjs/operators';
import * as ts from 'typescript';
import { createSourceFile$ } from '../tsSourceFile';

export const getInfo = (node: ts.Node) => {
  const file = node.getSourceFile();

  const { line, character: column } = file.getLineAndCharacterOfPosition(
    node.getStart(),
  );

  const fileName = file.fileName;
  const lineNumber = line + 1;
  const columnNumber = column + 1;

  return {
    fileName,
    lineNumber,
    columnNumber,
  };
};

export const findNode$ = <T extends ts.Node>(
  predicate: (node: ts.Node) => boolean,
) => (rootNode: ts.Node): Observable<T> => {
  return new Observable(subscriber => {
    let resolvedNode;

    function _traverse(node: ts.Node) {
      if (predicate(node)) {
        resolvedNode = node as T;
        subscriber.next(resolvedNode);
        subscriber.complete();
        return;
      }

      ts.forEachChild(node, _traverse);
    }

    _traverse(rootNode);

    subscriber.error(new Fault('Node not found', { rootNode }));
  });
};

export const findNodeAtCursor$ = <T extends ts.Node>(
  predicate: (node: ts.Node) => boolean = () => true,
) => (cursor: TagCursor) =>
  createSourceFile$(cursor.fileName).pipe(
    pluck('file'),
    switchMap(findNode$<T>(R.both(isAtCursor(cursor), predicate))),
    catchThrowFault('Node not found at cursor', { cursor }),
  );

export const findElementAtCursor$ = findNodeAtCursor$<ts.JsxElement>(
  ts.isJsxElement,
);

export const findTemplateStringAtCursor$ = findNodeAtCursor$<
  ts.TaggedTemplateExpression
>(ts.isTaggedTemplateExpression);

export const isAtCursor = (cursor: TagCursor) => (node: ts.Node) => {
  const { lineNumber, columnNumber } = getInfo(node);

  return (
    lineNumber === cursor.lineNumber && columnNumber === cursor.columnNumber
  );
};
