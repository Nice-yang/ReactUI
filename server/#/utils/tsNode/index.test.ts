import { createSourceFile$ } from '#/utils/tsSourceFile';
import { pluck, switchMap } from 'rxjs/operators';
import * as ts from 'typescript';
import { findNode$ } from '.';

describe('findNode', () => {
  const getRootNode$ = createSourceFile$(
    `${__dirname}/__fixtures__/findNode.tsx`,
  ).pipe(pluck('file'));

  it('success', done => {
    getRootNode$.pipe(switchMap(findNode$(ts.isJsxElement))).subscribe(node => {
      expect(node).toMatchSnapshot();
      expect(node.getText()).toMatchSnapshot();
      done();
    });
  });

  it('fail', done => {
    getRootNode$.pipe(switchMap(findNode$(ts.isJsxFragment))).subscribe({
      error: x => {
        expect(x).toMatchSnapshot();
        done();
      },
    });
  });
});
