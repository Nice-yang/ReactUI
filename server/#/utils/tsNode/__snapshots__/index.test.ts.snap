// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findNode fail 1`] = `
Object {
  "data": Object {
    "rootNode": SourceFileObject {
      "amdDependencies": Array [],
      "bindDiagnostics": Array [],
      "bindSuggestionDiagnostics": undefined,
      "checkJsDirective": undefined,
      "end": 136,
      "endOfFileToken": TokenObject {
        "end": 136,
        "flags": 0,
        "kind": 1,
        "modifierFlagsCache": 0,
        "parent": [Circular],
        "pos": 135,
        "transformFlags": 0,
      },
      "externalModuleIndicator": NodeObject {
        "decorators": undefined,
        "end": 31,
        "flags": 0,
        "importClause": NodeObject {
          "end": 17,
          "flags": 0,
          "kind": 250,
          "modifierFlagsCache": 0,
          "namedBindings": NodeObject {
            "end": 17,
            "flags": 0,
            "kind": 251,
            "modifierFlagsCache": 0,
            "name": IdentifierObject {
              "end": 17,
              "escapedText": "React",
              "flags": 0,
              "modifierFlagsCache": 0,
              "parent": [Circular],
              "pos": 11,
              "transformFlags": 0,
            },
            "parent": [Circular],
            "pos": 6,
            "transformFlags": 0,
          },
          "parent": [Circular],
          "pos": 6,
          "transformFlags": 0,
        },
        "kind": 249,
        "modifierFlagsCache": 536870912,
        "modifiers": undefined,
        "moduleSpecifier": TokenObject {
          "end": 30,
          "flags": 0,
          "kind": 10,
          "modifierFlagsCache": 0,
          "parent": [Circular],
          "pos": 22,
          "text": "react",
          "transformFlags": 0,
        },
        "parent": [Circular],
        "pos": 0,
        "transformFlags": 0,
      },
      "fileName": "/Users/I0459/scratch/personal/reactui/server/#/utils/tsNode/__fixtures__/findNode.tsx",
      "flags": 0,
      "hasNoDefaultLib": false,
      "identifierCount": 7,
      "identifiers": Map {
        "React" => "React",
        "react" => "react",
        "Component" => "Component",
        "div" => "div",
        "button" => "button",
      },
      "isDeclarationFile": false,
      "kind": 284,
      "languageVariant": 1,
      "languageVersion": 7,
      "libReferenceDirectives": Array [],
      "modifierFlagsCache": 0,
      "nodeCount": 31,
      "parent": undefined,
      "parseDiagnostics": Array [],
      "pos": 0,
      "pragmas": Map {},
      "referencedFiles": Array [],
      "scriptKind": 4,
      "statements": Array [
        NodeObject {
          "decorators": undefined,
          "end": 31,
          "flags": 0,
          "importClause": NodeObject {
            "end": 17,
            "flags": 0,
            "kind": 250,
            "modifierFlagsCache": 0,
            "namedBindings": NodeObject {
              "end": 17,
              "flags": 0,
              "kind": 251,
              "modifierFlagsCache": 0,
              "name": IdentifierObject {
                "end": 17,
                "escapedText": "React",
                "flags": 0,
                "modifierFlagsCache": 0,
                "parent": [Circular],
                "pos": 11,
                "transformFlags": 0,
              },
              "parent": [Circular],
              "pos": 6,
              "transformFlags": 0,
            },
            "parent": [Circular],
            "pos": 6,
            "transformFlags": 0,
          },
          "kind": 249,
          "modifierFlagsCache": 536870912,
          "modifiers": undefined,
          "moduleSpecifier": TokenObject {
            "end": 30,
            "flags": 0,
            "kind": 10,
            "modifierFlagsCache": 0,
            "parent": [Circular],
            "pos": 22,
            "text": "react",
            "transformFlags": 0,
          },
          "parent": [Circular],
          "pos": 0,
          "transformFlags": 0,
        },
        NodeObject {
          "declarationList": NodeObject {
            "declarations": Array [
              NodeObject {
                "end": 107,
                "flags": 0,
                "initializer": NodeObject {
                  "body": NodeObject {
                    "end": 107,
                    "expression": NodeObject {
                      "children": Array [
                        TokenObject {
                          "containsOnlyTriviaWhiteSpaces": true,
                          "end": 71,
                          "flags": 0,
                          "kind": 11,
                          "modifierFlagsCache": 0,
                          "parent": [Circular],
                          "pos": 66,
                          "text": "
    ",
                          "transformFlags": 0,
                        },
                        NodeObject {
                          "children": Array [
                            TokenObject {
                              "containsOnlyTriviaWhiteSpaces": false,
                              "end": 87,
                              "flags": 0,
                              "kind": 11,
                              "modifierFlagsCache": 0,
                              "parent": [Circular],
                              "pos": 79,
                              "text": "Click me",
                              "transformFlags": 0,
                            },
                          ],
                          "closingElement": NodeObject {
                            "end": 96,
                            "flags": 0,
                            "kind": 263,
                            "modifierFlagsCache": 0,
                            "parent": [Circular],
                            "pos": 87,
                            "tagName": IdentifierObject {
                              "end": 95,
                              "escapedText": "button",
                              "flags": 0,
                              "modifierFlagsCache": 0,
                              "parent": [Circular],
                              "pos": 89,
                              "transformFlags": 0,
                            },
                            "transformFlags": 0,
                          },
                          "end": 96,
                          "flags": 0,
                          "kind": 260,
                          "modifierFlagsCache": 0,
                          "openingElement": NodeObject {
                            "attributes": NodeObject {
                              "end": 78,
                              "flags": 0,
                              "kind": 268,
                              "modifierFlagsCache": 0,
                              "parent": [Circular],
                              "pos": 78,
                              "properties": Array [],
                              "transformFlags": 0,
                            },
                            "end": 79,
                            "flags": 0,
                            "kind": 262,
                            "modifierFlagsCache": 0,
                            "parent": [Circular],
                            "pos": 71,
                            "tagName": IdentifierObject {
                              "end": 78,
                              "escapedText": "button",
                              "flags": 0,
                              "modifierFlagsCache": 0,
                              "parent": [Circular],
                              "pos": 72,
                              "transformFlags": 0,
                            },
                            "transformFlags": 0,
                            "typeArguments": undefined,
                          },
                          "parent": [Circular],
                          "pos": 71,
                          "transformFlags": 0,
                        },
                        TokenObject {
                          "containsOnlyTriviaWhiteSpaces": true,
                          "end": 99,
                          "flags": 0,
                          "kind": 11,
                          "modifierFlagsCache": 0,
                          "parent": [Circular],
                          "pos": 96,
                          "text": "
  ",
                          "transformFlags": 0,
                        },
                      ],
                      "closingElement": NodeObject {
                        "end": 105,
                        "flags": 0,
                        "kind": 263,
                        "modifierFlagsCache": 0,
                        "parent": [Circular],
                        "pos": 99,
                        "tagName": IdentifierObject {
                          "end": 104,
                          "escapedText": "div",
                          "flags": 0,
                          "modifierFlagsCache": 0,
                          "parent": [Circular],
                          "pos": 101,
                          "transformFlags": 0,
                        },
                        "transformFlags": 0,
                      },
                      "end": 105,
                      "flags": 0,
                      "kind": 260,
                      "modifierFlagsCache": 0,
                      "openingElement": NodeObject {
                        "attributes": NodeObject {
                          "end": 65,
                          "flags": 0,
                          "kind": 268,
                          "modifierFlagsCache": 0,
                          "parent": [Circular],
                          "pos": 65,
                          "properties": Array [],
                          "transformFlags": 0,
                        },
                        "end": 66,
                        "flags": 0,
                        "kind": 262,
                        "modifierFlagsCache": 0,
                        "parent": [Circular],
                        "pos": 58,
                        "tagName": IdentifierObject {
                          "end": 65,
                          "escapedText": "div",
                          "flags": 0,
                          "modifierFlagsCache": 0,
                          "parent": [Circular],
                          "pos": 62,
                          "transformFlags": 0,
                        },
                        "transformFlags": 0,
                        "typeArguments": undefined,
                      },
                      "parent": [Circular],
                      "pos": 58,
                      "transformFlags": 0,
                    },
                    "flags": 0,
                    "kind": 195,
                    "modifierFlagsCache": 0,
                    "parent": [Circular],
                    "pos": 56,
                    "transformFlags": 0,
                  },
                  "end": 107,
                  "equalsGreaterThanToken": TokenObject {
                    "end": 56,
                    "flags": 0,
                    "kind": 37,
                    "modifierFlagsCache": 0,
                    "parent": [Circular],
                    "pos": 53,
                    "transformFlags": 0,
                  },
                  "flags": 0,
                  "kind": 197,
                  "modifierFlagsCache": 536870912,
                  "modifiers": undefined,
                  "parameters": Array [],
                  "parent": [Circular],
                  "pos": 50,
                  "transformFlags": 0,
                  "typeParameters": undefined,
                },
                "kind": 237,
                "modifierFlagsCache": 0,
                "name": IdentifierObject {
                  "end": 48,
                  "escapedText": "Component",
                  "flags": 0,
                  "modifierFlagsCache": 0,
                  "parent": [Circular],
                  "pos": 38,
                  "transformFlags": 0,
                },
                "parent": [Circular],
                "pos": 38,
                "transformFlags": 0,
                "type": undefined,
              },
            ],
            "end": 107,
            "flags": 2,
            "kind": 238,
            "modifierFlagsCache": 0,
            "parent": [Circular],
            "pos": 31,
            "transformFlags": 0,
          },
          "decorators": undefined,
          "end": 108,
          "flags": 0,
          "kind": 219,
          "modifierFlagsCache": 0,
          "modifiers": undefined,
          "parent": [Circular],
          "pos": 31,
          "transformFlags": 0,
        },
        NodeObject {
          "decorators": undefined,
          "end": 135,
          "expression": IdentifierObject {
            "end": 134,
            "escapedText": "Component",
            "flags": 0,
            "modifierFlagsCache": 0,
            "parent": [Circular],
            "pos": 124,
            "transformFlags": 0,
          },
          "flags": 0,
          "kind": 254,
          "modifierFlagsCache": 0,
          "modifiers": undefined,
          "parent": [Circular],
          "pos": 108,
          "transformFlags": 0,
        },
      ],
      "text": "import * as React from 'react';

const Component = () => (
  <div>
    <button>Click me</button>
  </div>
);

export default Component;
",
      "transformFlags": 0,
      "typeReferenceDirectives": Array [],
    },
  },
  "level": 0,
  "message": "Node not found",
  "stack": undefined,
}
`;

exports[`findNode success 1`] = `
NodeObject {
  "children": Array [
    TokenObject {
      "containsOnlyTriviaWhiteSpaces": true,
      "end": 71,
      "flags": 0,
      "kind": 11,
      "modifierFlagsCache": 0,
      "parent": [Circular],
      "pos": 66,
      "text": "
    ",
      "transformFlags": 0,
    },
    NodeObject {
      "children": Array [
        TokenObject {
          "containsOnlyTriviaWhiteSpaces": false,
          "end": 87,
          "flags": 0,
          "kind": 11,
          "modifierFlagsCache": 0,
          "parent": [Circular],
          "pos": 79,
          "text": "Click me",
          "transformFlags": 0,
        },
      ],
      "closingElement": NodeObject {
        "end": 96,
        "flags": 0,
        "kind": 263,
        "modifierFlagsCache": 0,
        "parent": [Circular],
        "pos": 87,
        "tagName": IdentifierObject {
          "end": 95,
          "escapedText": "button",
          "flags": 0,
          "modifierFlagsCache": 0,
          "parent": [Circular],
          "pos": 89,
          "transformFlags": 0,
        },
        "transformFlags": 0,
      },
      "end": 96,
      "flags": 0,
      "kind": 260,
      "modifierFlagsCache": 0,
      "openingElement": NodeObject {
        "attributes": NodeObject {
          "end": 78,
          "flags": 0,
          "kind": 268,
          "modifierFlagsCache": 0,
          "parent": [Circular],
          "pos": 78,
          "properties": Array [],
          "transformFlags": 0,
        },
        "end": 79,
        "flags": 0,
        "kind": 262,
        "modifierFlagsCache": 0,
        "parent": [Circular],
        "pos": 71,
        "tagName": IdentifierObject {
          "end": 78,
          "escapedText": "button",
          "flags": 0,
          "modifierFlagsCache": 0,
          "parent": [Circular],
          "pos": 72,
          "transformFlags": 0,
        },
        "transformFlags": 0,
        "typeArguments": undefined,
      },
      "parent": [Circular],
      "pos": 71,
      "transformFlags": 0,
    },
    TokenObject {
      "containsOnlyTriviaWhiteSpaces": true,
      "end": 99,
      "flags": 0,
      "kind": 11,
      "modifierFlagsCache": 0,
      "parent": [Circular],
      "pos": 96,
      "text": "
  ",
      "transformFlags": 0,
    },
  ],
  "closingElement": NodeObject {
    "end": 105,
    "flags": 0,
    "kind": 263,
    "modifierFlagsCache": 0,
    "parent": [Circular],
    "pos": 99,
    "tagName": IdentifierObject {
      "end": 104,
      "escapedText": "div",
      "flags": 0,
      "modifierFlagsCache": 0,
      "parent": [Circular],
      "pos": 101,
      "transformFlags": 0,
    },
    "transformFlags": 0,
  },
  "end": 105,
  "flags": 0,
  "kind": 260,
  "modifierFlagsCache": 0,
  "openingElement": NodeObject {
    "attributes": NodeObject {
      "end": 65,
      "flags": 0,
      "kind": 268,
      "modifierFlagsCache": 0,
      "parent": [Circular],
      "pos": 65,
      "properties": Array [],
      "transformFlags": 0,
    },
    "end": 66,
    "flags": 0,
    "kind": 262,
    "modifierFlagsCache": 0,
    "parent": [Circular],
    "pos": 58,
    "tagName": IdentifierObject {
      "end": 65,
      "escapedText": "div",
      "flags": 0,
      "modifierFlagsCache": 0,
      "parent": [Circular],
      "pos": 62,
      "transformFlags": 0,
    },
    "transformFlags": 0,
    "typeArguments": undefined,
  },
  "parent": NodeObject {
    "end": 107,
    "expression": [Circular],
    "flags": 0,
    "kind": 195,
    "modifierFlagsCache": 0,
    "parent": NodeObject {
      "body": [Circular],
      "end": 107,
      "equalsGreaterThanToken": TokenObject {
        "end": 56,
        "flags": 0,
        "kind": 37,
        "modifierFlagsCache": 0,
        "parent": [Circular],
        "pos": 53,
        "transformFlags": 0,
      },
      "flags": 0,
      "kind": 197,
      "modifierFlagsCache": 536870912,
      "modifiers": undefined,
      "parameters": Array [],
      "parent": NodeObject {
        "end": 107,
        "flags": 0,
        "initializer": [Circular],
        "kind": 237,
        "modifierFlagsCache": 0,
        "name": IdentifierObject {
          "end": 48,
          "escapedText": "Component",
          "flags": 0,
          "modifierFlagsCache": 0,
          "parent": [Circular],
          "pos": 38,
          "transformFlags": 0,
        },
        "parent": NodeObject {
          "declarations": Array [
            [Circular],
          ],
          "end": 107,
          "flags": 2,
          "kind": 238,
          "modifierFlagsCache": 0,
          "parent": NodeObject {
            "declarationList": [Circular],
            "decorators": undefined,
            "end": 108,
            "flags": 0,
            "kind": 219,
            "modifierFlagsCache": 0,
            "modifiers": undefined,
            "parent": SourceFileObject {
              "amdDependencies": Array [],
              "bindDiagnostics": Array [],
              "bindSuggestionDiagnostics": undefined,
              "checkJsDirective": undefined,
              "end": 136,
              "endOfFileToken": TokenObject {
                "end": 136,
                "flags": 0,
                "kind": 1,
                "modifierFlagsCache": 0,
                "parent": [Circular],
                "pos": 135,
                "transformFlags": 0,
              },
              "externalModuleIndicator": NodeObject {
                "decorators": undefined,
                "end": 31,
                "flags": 0,
                "importClause": NodeObject {
                  "end": 17,
                  "flags": 0,
                  "kind": 250,
                  "modifierFlagsCache": 0,
                  "namedBindings": NodeObject {
                    "end": 17,
                    "flags": 0,
                    "kind": 251,
                    "modifierFlagsCache": 0,
                    "name": IdentifierObject {
                      "end": 17,
                      "escapedText": "React",
                      "flags": 0,
                      "modifierFlagsCache": 0,
                      "parent": [Circular],
                      "pos": 11,
                      "transformFlags": 0,
                    },
                    "parent": [Circular],
                    "pos": 6,
                    "transformFlags": 0,
                  },
                  "parent": [Circular],
                  "pos": 6,
                  "transformFlags": 0,
                },
                "kind": 249,
                "modifierFlagsCache": 536870912,
                "modifiers": undefined,
                "moduleSpecifier": TokenObject {
                  "end": 30,
                  "flags": 0,
                  "kind": 10,
                  "modifierFlagsCache": 0,
                  "parent": [Circular],
                  "pos": 22,
                  "text": "react",
                  "transformFlags": 0,
                },
                "parent": [Circular],
                "pos": 0,
                "transformFlags": 0,
              },
              "fileName": "/Users/I0459/scratch/personal/reactui/server/#/utils/tsNode/__fixtures__/findNode.tsx",
              "flags": 0,
              "hasNoDefaultLib": false,
              "identifierCount": 7,
              "identifiers": Map {
                "React" => "React",
                "react" => "react",
                "Component" => "Component",
                "div" => "div",
                "button" => "button",
              },
              "isDeclarationFile": false,
              "kind": 284,
              "languageVariant": 1,
              "languageVersion": 7,
              "libReferenceDirectives": Array [],
              "modifierFlagsCache": 0,
              "nodeCount": 31,
              "parent": undefined,
              "parseDiagnostics": Array [],
              "pos": 0,
              "pragmas": Map {},
              "referencedFiles": Array [],
              "scriptKind": 4,
              "statements": Array [
                NodeObject {
                  "decorators": undefined,
                  "end": 31,
                  "flags": 0,
                  "importClause": NodeObject {
                    "end": 17,
                    "flags": 0,
                    "kind": 250,
                    "modifierFlagsCache": 0,
                    "namedBindings": NodeObject {
                      "end": 17,
                      "flags": 0,
                      "kind": 251,
                      "modifierFlagsCache": 0,
                      "name": IdentifierObject {
                        "end": 17,
                        "escapedText": "React",
                        "flags": 0,
                        "modifierFlagsCache": 0,
                        "parent": [Circular],
                        "pos": 11,
                        "transformFlags": 0,
                      },
                      "parent": [Circular],
                      "pos": 6,
                      "transformFlags": 0,
                    },
                    "parent": [Circular],
                    "pos": 6,
                    "transformFlags": 0,
                  },
                  "kind": 249,
                  "modifierFlagsCache": 536870912,
                  "modifiers": undefined,
                  "moduleSpecifier": TokenObject {
                    "end": 30,
                    "flags": 0,
                    "kind": 10,
                    "modifierFlagsCache": 0,
                    "parent": [Circular],
                    "pos": 22,
                    "text": "react",
                    "transformFlags": 0,
                  },
                  "parent": [Circular],
                  "pos": 0,
                  "transformFlags": 0,
                },
                [Circular],
                NodeObject {
                  "decorators": undefined,
                  "end": 135,
                  "expression": IdentifierObject {
                    "end": 134,
                    "escapedText": "Component",
                    "flags": 0,
                    "modifierFlagsCache": 0,
                    "parent": [Circular],
                    "pos": 124,
                    "transformFlags": 0,
                  },
                  "flags": 0,
                  "kind": 254,
                  "modifierFlagsCache": 0,
                  "modifiers": undefined,
                  "parent": [Circular],
                  "pos": 108,
                  "transformFlags": 0,
                },
              ],
              "text": "import * as React from 'react';

const Component = () => (
  <div>
    <button>Click me</button>
  </div>
);

export default Component;
",
              "transformFlags": 0,
              "typeReferenceDirectives": Array [],
            },
            "pos": 31,
            "transformFlags": 0,
          },
          "pos": 31,
          "transformFlags": 0,
        },
        "pos": 38,
        "transformFlags": 0,
        "type": undefined,
      },
      "pos": 50,
      "transformFlags": 0,
      "typeParameters": undefined,
    },
    "pos": 56,
    "transformFlags": 0,
  },
  "pos": 58,
  "transformFlags": 0,
}
`;

exports[`findNode success 2`] = `
"<div>
    <button>Click me</button>
  </div>"
`;
